libaleth-interpreter/VM.cpp

```diff
@@ -380,17 +380,21 @@ void VM::interpretCases()
             m_runGas = m_rev >= EVMC_TANGERINE_WHISTLE ? 5000 : 0;
             evmc_address destination = toEvmC(asAddress(m_SP[0]));
 
-            // After EIP158 zero-value suicides do not have to pay account creation gas.
-            u256 const balance =
-                fromEvmC(m_context->host->get_balance(m_context, &m_message->destination));
-            if (balance > 0 || m_rev < EVMC_SPURIOUS_DRAGON)
+            // Self-destructs only have gas cost starting with Tangerine Whistle
+            if (m_rev >= EVMC_TANGERINE_WHISTLE)
             {
-                // After EIP150 hard fork charge additional cost of sending
-                // ethers to non-existing account.
-                int destinationExists =
-                    m_context->host->account_exists(m_context, &destination);
-                if (m_rev >= EVMC_TANGERINE_WHISTLE && !destinationExists)
-                    m_runGas += VMSchedule::callNewAccount;
+                // After EIP158 zero-value suicides do not have to pay account creation gas.
+                u256 const balance =
+                    fromEvmC(m_context->host->get_balance(m_context, &m_message->destination));
```

Comment:
This is still "wrong", because `get_balance()` will be executed in case of `m_rev == EVMC_TANGERINE_WHISTLE` where its value does not matter.

Reply:
@chfast Ah thought that the changes should apply to Homestead / Frontier, I'll take another look at which hard forks the EIPs were enabled on. According to this they were both enabled on Tangerine Whistle but that's not right if we don't care about the balance on Tangerine Whistle: https://ethereum.stackexchange.com/questions/13014/please-provide-a-summary-of-the-ethereum-hard-forks

Reply:
No need to check external sources. The logic here is correct. Just if you check the `if` at https://github.com/ethereum/aleth/pull/5718/files#diff-4a2862de6e660dc5045835d8879be719R389 you will see that:
1. you can swap the checks in `||`.
2. The balance has be checked only if the first condition is false.

Reply:
@chfast I'm confused...the line that you're referring to comes after balance retrieval so changing the checks will have no impact on whether or not we check the balance on Tangerine Whistle? 

Line in question: https://github.com/ethereum/aleth/blob/fcb1f3d4d953896a4114a6cd9e77f98826437926/libaleth-interpreter/VM.cpp#L389

Reply:
@chfast I'm still a little confused about this part of the aleth-interpreter changes: https://github.com/ethereum/aleth/blob/29ed24eddb9e40cc450259916c7892f0685fa8e0/libaleth-interpreter/VM.cpp#L383-L388

The check for EVMC_TANGERINE_WHISTLE on line 387 implies that on that hard fork, all suicides (0-value or otherwise) have an account-creation charge. This implies that EIP158 (which removes the account-creation charge for 0-value selfdestructs) was included in Spurious Dragon, but the following says that it was included in Tangerine Whistle: https://ethereum.stackexchange.com/questions/13014/please-provide-a-summary-of-the-ethereum-hard-forks

So to summarize, EIP158 was included in Spurious Dragon and not Tangerine Whistle?

Reply:
EIP-158 was activated at block 2'675'000 which is Spurious Dragon https://github.com/ethereum/aleth/blob/7e6095dc3d6c9cefac4b0867f60d06f955ab7c7e/libethashseal/genesis/mainNetwork.cpp#L29

in EIPs actually EIP-161 supercedes EIP-158 and is listed in meta-EIP https://eips.ethereum.org/EIPS/eip-607

Reply:
@gumb0 Thanks for the info! Is there a reliable list that you can recommend of hard forks + the block numbers that they occurred at?

Reply:
I think meta-EIPs should correctly reflect the reality of mainnet https://eips.ethereum.org/meta